### ISSUE TITLE
Sway hard freeze, z-fighting and framebuffer corruption on nouveau

### ISSUE THREAD

**Issue by 4censord** [swaywm/sway#5889]
### Please fill out the following:
Version:
sway 1.5.1
wlroots 0.12.0
xwayland 1.20.0.r797.g5d73a8b59
- **Debug Log:**
  [debug-log](https://gist.github.com/dsko1234/b965dba1c1afeed783a52238c28d5495)

- **Configuration File:**
Its reproduce-able with only my output config added to the default configuration.
```
output eDP-1 resolution 3840x2160 position 10000 10000

set $laptop eDP-1
bindswitch --reload --locked lid:on output $laptop disable
bindswitch --reload --locked lid:off output $laptop enable

output 'NEC Corporation EX231W 45209021EB'            resolution 1920x1080 position  0  0  transform 270
output 'Iiyama North America PL2792UH 1176903301664'  resolution 3840x2160 position  1080  0 
output 'Iiyama North America PL2792H  1176603602576'  resolution 1920x1080 position  4920  0 
```

- **Stack Trace:**
  sway doesn't crash but hangs indefinitely

### Description
My setup 
Laptop Thinkpad P50, nvidia Quadro M2000M, nouveau

Multiple screens, arranged from left to right
Laptop Screen 3840x2160  
1st Screen 1920x1080 minidisplay port  
2nd Screen 3840x2160 thunderbolt dock  
3rd Screen 1920x1080 thunderbolt dock  

I have had multiple different types of error.
Most common: Sway just freezes, might be brought back with sysrq-keys
Kinda common: something looking like z-fighting, most of the time on one of the 4k monitors
Kinda rare: Framebuffer corruption, see picture

### Reproduction: freeze

start sway  
start Firefox (xwayland)  
scroll up and down  
sway freezes  

its also reproducible with other Xwayland programs, its just easiest with Firefox
if the laptop screen is disabled it also happens, just not as fast/easy to reproduce. 

### Reproduction: framebuffer corruption:

Sometimes starting [Event Horizon](https://store.steampowered.com/app/465000/Event_Horizon/) via steam looks like it corrupts the framebuffer.
Sometimes sway and all other wayland programms are fine but most of the time everything looks corrupt

### Reproduction: z-fighting

happens seemingly random, but more often if the laptopscreen is enabled

**Comment by emersion** [comment]
Can you have a look at dmesg?

**Comment by dsko1234** [comment]
I have taken a look at yesterdays [dmesg](https://gist.github.com/dsko1234/520e1a68e2ed4808fafb1c429bef87e5)
for some reason i can not directly attach files to issues.

for the hang I reported is from line 4120 to 4347

for reference: `lspci | grep VGA`
`01:00.0 VGA compatible controller: NVIDIA Corporation GM107GLM [Quadro M2000M] (rev a2)`
so my gpu is at 0000:01:00.0

the only thing that was marked red about nouveau
```
Dec 15 22:09:43 thearch kernel: nouveau 0000:01:00.0: bus: MMIO read of 00000000 FAULT at 022554 [ IBUS ]
Dec 15 22:09:43 thearch kernel: DMAR: DRHD: handling fault status reg 2
Dec 15 22:09:43 thearch kernel: DMAR: [DMA Read] Request device [01:00.0] PASID ffffffff fault addr 0 [fault reason 06] PTE Read access is not set
Dec 15 22:09:43 thearch kernel: nouveau 0000:01:00.0: bus: MMIO read of 00000000 FAULT at 10ac08 [ IBUS ]
```

**Comment by dsko1234** [comment]
It just froze again, on opening spotify.
3 xwayland windows open,
scratchpad: alfaview (qt application without wayland support)
2nd screen, floating, spotify
3rd screen, right half, xeyes 
laptopscreen was disabled by closing the lid 

I ran spotify via `flatpack run com.spotify.client`, and sway froze.

- [swaylog](https://gist.github.com/dsko1234/aab5455040473098623270603d43576c#file-swaylog)
- [Dmesg](https://gist.github.com/dsko1234/aab5455040473098623270603d43576c#file-dmesg-2020-12-16-1000) my session starts at line 3151

**Comment by dsko1234** [comment]
If i let it hanging for long enough it crashes.
i got a coredump 
```
# coredumpctl gdb sway
> bt full
Reading symbols from /usr/bin/sway...
(No debugging symbols found in /usr/bin/sway)

warning: Can't open file /memfd:gdk-wayland (deleted) during file-backed mapping note processing

warning: Can't open file /memfd:wayland-shm (deleted) during file-backed mapping note processing

warning: Can't open file /memfd:wayland-cursor (deleted) during file-backed mapping note processing

warning: Can't open file /dev/shm/wayland.mozilla.ipc.0 (deleted) during file-backed mapping note processing

warning: Can't open file /dev/shm/mako-mlgbkc (deleted) during file-backed mapping note processing

warning: Can't open file /dev/shm/wayland.mozilla.ipc.1 (deleted) during file-backed mapping note processing

warning: Can't open file /run/user/1000/sway-client-LjEawM (deleted) during file-backed mapping note processing

warning: Can't open file /run/user/1000/sway-client-ULQBRH (deleted) during file-backed mapping note processing
[dmesg-cordump.txt](https://github.com/swaywm/sway/files/5704198/dmesg-cordump.txt)

warning: Can't open file /memfd:xwayland-shared (deleted) during file-backed mapping note processing

warning: Can't open file /run/user/1000/sway-client-QydC9J (deleted) during file-backed mapping note processing
[New LWP 743]
[New LWP 757]
[New LWP 758]
[New LWP 759]
[New LWP 756]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".
Core was generated by `sway -d'.
Program terminated with signal SIGABRT, Aborted.
#0  0x00007f2fa08c2615 in raise () from /usr/lib/libc.so.6
[Current thread is 1 (Thread 0x7f2f9fc73940 (LWP 743))]
(gdb) bt full
#0  0x00007f2fa08c2615 in raise () at /usr/lib/libc.so.6
#1  0x00007f2fa08ab862 in abort () at /usr/lib/libc.so.6
#2  0x00007f2fa08ab747 in _nl_load_domain.cold () at /usr/lib/libc.so.6
#3  0x00007f2fa08babf6 in  () at /usr/lib/libc.so.6
#4  0x00007f2f989bf7a0 in  () at /usr/lib/libdrm_nouveau.so.2
#5  0x00007f2f989bf703 in nouveau_pushbuf_data () at /usr/lib/libdrm_nouveau.so.2
#6  0x00007f2f989bf820 in  () at /usr/lib/libdrm_nouveau.so.2
#7  0x00007f2f989bfc47 in  () at /usr/lib/libdrm_nouveau.so.2
#8  0x00007f2f989c099a in  () at /usr/lib/libdrm_nouveau.so.2
#9  0x00007f2f9ee54985 in  () at /usr/lib/dri/nouveau_dri.so
#10 0x00007f2f9ed9966a in  () at /usr/lib/dri/nouveau_dri.so
#11 0x00007f2f9e5fc4f2 in  () at /usr/lib/dri/nouveau_dri.so
#12 0x00007f2fa0af2e41 in  () at /usr/lib/libwlroots.so.7
#13 0x00007f2fa0afef2a in  () at /usr/lib/libwlroots.so.7
#14 0x00007f2fa0b3194c in  () at /usr/lib/libwlroots.so.7
#15 0x00007f2fa0b33945 in  () at /usr/lib/libwlroots.so.7
#16 0x00007f2fa0b20a12 in wlr_cursor_set_surface () at /usr/lib/libwlroots.so.7
#17 0x00007f2fa0b3c70e in  () at /usr/lib/libwlroots.so.7
#18 0x00007f2fa0b16f92 in  () at /usr/lib/libwlroots.so.7
#19 0x00007f2fa027dacd in  () at /usr/lib/libffi.so.7
#20 0x00007f2fa027d03a in  () at /usr/lib/libffi.so.7
#21 0x00007f2fa0b88f62 in  () at /usr/lib/libwayland-server.so.0
#22 0x00007f2fa0b852dc in  () at /usr/lib/libwayland-server.so.0
#23 0x00007f2fa0b86faa in wl_event_loop_dispatch () at /usr/lib/libwayland-server.so.0
#24 0x00007f2fa0b854e7 in wl_display_run () at /usr/lib/libwayland-server.so.0
#25 0x00005584c154282f in  ()
#26 0x00007f2fa08ad152 in __libc_start_main () at /usr/lib/libc.so.6
#27 0x00005584c1542bae in  ()

```
the swaylog
[swaylog.txt](https://github.com/swaywm/sway/files/5704206/swaylog.txt)
the dmesg entry for the crash
[dmesg-cordump.txt](https://github.com/swaywm/sway/files/5704208/dmesg-cordump.txt)

**Comment by emersion** [comment]
Looks like a bug in the nouveau Mesa driver. Please report it to Mesa.

**Comment by dsko1234** [comment]
will do

**Comment by emersion** [comment]
Sounds similar to https://github.com/swaywm/sway/issues/5883

**Comment by dsko1234** [comment]
i will try if i can reproduce by their method and report back

**Comment by aragon999** [comment]
I am not sure, I wanted to look more into this (in order to find out where to report), I stumbeled upon this issue and #5883 so I wanted to add my observations, hopefully this is useful.

So this issue happens quite some time for me, but I had no time to investigate this issue further.

So additionally to the original posted sway.log I also get the following message:
```
sway: ../libdrm-2.4.102/nouveau/pushbuf.c:723: nouveau_pushbuf_data: Assertion `kref' failed.
```
See the corresponding libdrm source: https://gitlab.freedesktop.org/mesa/drm/-/blob/master/nouveau/pushbuf.c#L723

At the same time I find in my system log the following:
```
Dez 21 17:14:07 nvidia-desktop kernel: nouveau 0000:01:00.0: gr: TRAP ch 2 [017f83e000 systemd-logind[423]]
Dez 21 17:14:07 nvidia-desktop kernel: nouveau 0000:01:00.0: gr: GPC0/TPC0/TEX: 80000041
Dez 21 17:14:07 nvidia-desktop kernel: nouveau 0000:01:00.0: gr: GPC0/TPC1/TEX: 80000041
Dez 21 17:14:07 nvidia-desktop kernel: nouveau 0000:01:00.0: gr: GPC0/TPC2/TEX: 80000041
Dez 21 17:14:07 nvidia-desktop kernel: nouveau 0000:01:00.0: gr: GPC0/TPC3/TEX: 80000041
Dez 21 17:14:07 nvidia-desktop kernel: nouveau 0000:01:00.0: gr: GPC0/TPC4/TEX: 80000041
Dez 21 17:14:07 nvidia-desktop kernel: nouveau 0000:01:00.0: gr: GPC1/TPC0/TEX: 80000041
Dez 21 17:14:07 nvidia-desktop kernel: nouveau 0000:01:00.0: gr: GPC1/TPC1/TEX: 80000041
Dez 21 17:14:07 nvidia-desktop kernel: nouveau 0000:01:00.0: gr: GPC1/TPC2/TEX: 80000041
Dez 21 17:14:07 nvidia-desktop kernel: nouveau 0000:01:00.0: gr: GPC1/TPC3/TEX: 80000041
Dez 21 17:14:07 nvidia-desktop kernel: nouveau 0000:01:00.0: gr: GPC1/TPC4/TEX: 80000041
Dez 21 17:14:07 nvidia-desktop kernel: nouveau 0000:01:00.0: fifo: fault 00 [READ] at 0000000010c82000 engine 00 [GR] client 1e [GPC1/PE_7] reason 00 [PDE] on channel 2 [017f83e000 systemd-logind[423]]
Dez 21 17:14:07 nvidia-desktop kernel: nouveau 0000:01:00.0: fifo: channel 2: killed
Dez 21 17:14:07 nvidia-desktop kernel: nouveau 0000:01:00.0: fifo: runlist 0: scheduled for recovery
Dez 21 17:14:07 nvidia-desktop kernel: nouveau 0000:01:00.0: fifo: engine 0: scheduled for recovery
Dez 21 17:14:07 nvidia-desktop kernel: nouveau 0000:01:00.0: fifo: engine 7: scheduled for recovery
Dez 21 17:14:07 nvidia-desktop kernel: nouveau 0000:01:00.0: systemd-logind[423]: channel 2 killed!
```

So my current understanding of the bug is, that something from sway is triggering a bug in Nouveau which then leads to Nouveau trying to restart and at the same time sway/wlroots wants to do something but that fails since the Nouveau is trying to restart (never saw it actually restarting). Please correct me if I have a wrong argumentation or if there is a bug in my thoughts.

My full logs: https://gist.github.com/aragon999/759dd59ce6b1301440f23d0c184c8542

**Comment by BenMueller** [comment]
I can second that this is an upstream mesa/nouveau issue, at least on the P50 with M2000M. While it's basically a quadro variant of the 960m (which is supported) the variant is so rare it seems it never got added upstream. There was a RHEL bugzilla report I used to have pinned about this issue (probably still out there) but since I switched out my P50 for a dell machine thanks to the issue, I no longer have it handy since it hadn't been touched in years.
